### ISSUE TITLE
Webhooks unreliable - Often run into a database lock

### ISSUE THREAD

**Issue by mrnugget** [applikatoni/applikatoni#35]
While testing around locally I ran into an issue with the webhooks, implemented in #31. It seems like the webhooks run into database locks really, really often. I'm still not sure why exactly but I know "how" this is happening: the webhook notifier loads a deployment from the database on _every_ log message sent through the system. And most of these `SELECT` statements fail. Just to give you an idea of how often this happens, here is the log output:

```
2016/01/19 07:51:53 STARTING DEPLOYMENT: deployment_id=235
::1 - - [19/Jan/2016:07:51:52 +0100] "POST /super-duper-app/deployments HTTP/1.1" 303 0
::1 - - [19/Jan/2016:07:51:53 +0100] "GET /super-duper-app/deployments/235 HTTP/1.1" 200 10689
2016/01/19 07:51:53 STARTING STAGE: CHECK_CONNECTION
2016/01/19 07:51:54 2.web.shipping-company.com:22 -- STARTING: test -d /var/www/shipping-company.com
2016/01/19 07:51:54 1.web.shipping-company.com:22 -- STARTING: test -d /var/www/shipping-company.com
2016/01/19 07:51:54 1.workers.shipping-company.com:22 -- STARTING: test -d /var/www/shipping-company.com && cd /var/www/shipping-company.com/current
2016/01/19 07:51:54 1.web.shipping-company.com:22 -- SUCCESS: "test -d /var/www/shipping-company.com"
2016/01/19 07:51:54 1.workers.shipping-company.com:22 -- SUCCESS: "test -d /var/www/shipping-company.com && cd /var/www/shipping-company.com/current"
2016/01/19 07:51:54 2.web.shipping-company.com:22 -- SUCCESS: "test -d /var/www/shipping-company.com"
2016/01/19 07:51:54 STAGE RESULT: 1.web.shipping-company.com:22 - execution of stage CHECK_CONNECTION successful (144.395246ms)
2016/01/19 07:51:54 STAGE RESULT: 1.workers.shipping-company.com:22 - execution of stage CHECK_CONNECTION successful (293.338033ms)
2016/01/19 07:51:55 STAGE RESULT: 2.web.shipping-company.com:22 - execution of stage CHECK_CONNECTION successful (505.786888ms)
2016/01/19 07:51:56 STAGE SUCCESS: CHECK_CONNECTION
2016/01/19 07:51:56 STARTING STAGE: PRE_DEPLOYMENT
2016/01/19 07:51:56 2.web.shipping-company.com:22 -- STARTING: cd /var/www/shipping-company.com && pwd
2016/01/19 07:51:56 1.web.shipping-company.com:22 -- STARTING: cd /var/www/shipping-company.com && pwd
2016/01/19 07:51:56 1.workers.shipping-company.com:22 -- STARTING: echo 'shutting down workers'
2016/01/19 07:51:56 1.web.shipping-company.com:22 -- STDOUT -- /var/www/shipping-company.com
2016/01/19 07:51:56 1.web.shipping-company.com:22 -- SUCCESS: "cd /var/www/shipping-company.com && pwd"
2016/01/19 07:51:56 2.web.shipping-company.com:22 -- STDOUT -- /var/www/shipping-company.com
2016/01/19 07:51:56 2.web.shipping-company.com:22 -- SUCCESS: "cd /var/www/shipping-company.com && pwd"
2016/01/19 07:51:56 1.workers.shipping-company.com:22 -- STDOUT -- shutting down workers
2016/01/19 07:51:57 1.workers.shipping-company.com:22 -- SUCCESS: "echo 'shutting down workers'"
2016/01/19 07:51:58 STAGE RESULT: 1.web.shipping-company.com:22 - execution of stage PRE_DEPLOYMENT successful (128.502005ms)
2016/01/19 07:51:59 Could not find deployment with id 235, database is locked
2016/01/19 07:51:59 Could not find deployment with id 235, database is locked
2016/01/19 07:51:59 Could not find deployment with id 235, database is locked
2016/01/19 07:51:59 Could not find deployment with id 235, database is locked
2016/01/19 07:51:59 Could not find deployment with id 235, database is locked
2016/01/19 07:51:59 Could not find deployment with id 235, database is locked
2016/01/19 07:51:59 STAGE RESULT: 2.web.shipping-company.com:22 - execution of stage PRE_DEPLOYMENT successful (130.945111ms)
2016/01/19 07:51:59 STAGE RESULT: 1.workers.shipping-company.com:22 - execution of stage PRE_DEPLOYMENT successful (132.300165ms)
2016/01/19 07:51:59 STAGE SUCCESS: PRE_DEPLOYMENT
2016/01/19 07:51:59 STARTING STAGE: CODE_DEPLOYMENT
2016/01/19 07:51:59 2.web.shipping-company.com:22 -- STARTING: echo 'deploying code and sleeping' && sleep 20
2016/01/19 07:51:59 1.web.shipping-company.com:22 -- STARTING: echo 'deploying code and sleeping' && sleep 20
2016/01/19 07:51:59 1.workers.shipping-company.com:22 -- STARTING: echo 'deploying code and sleeping' && sleep 20
2016/01/19 07:51:59 1.workers.shipping-company.com:22 -- STDOUT -- deploying code and sleeping
2016/01/19 07:51:59 1.web.shipping-company.com:22 -- STDOUT -- deploying code and sleeping
2016/01/19 07:51:59 2.web.shipping-company.com:22 -- STDOUT -- deploying code and sleeping
2016/01/19 07:52:00 Could not find deployment with id 235, database is locked
2016/01/19 07:52:14 1.workers.shipping-company.com:22 -- SUCCESS: "echo 'deploying code and sleeping' && sleep 20"
2016/01/19 07:52:14 2.web.shipping-company.com:22 -- SUCCESS: "echo 'deploying code and sleeping' && sleep 20"
2016/01/19 07:52:14 1.web.shipping-company.com:22 -- SUCCESS: "echo 'deploying code and sleeping' && sleep 20"
2016/01/19 07:52:14 STAGE RESULT: 1.workers.shipping-company.com:22 - execution of stage CODE_DEPLOYMENT successful (20.124435914s)
2016/01/19 07:52:14 STAGE RESULT: 2.web.shipping-company.com:22 - execution of stage CODE_DEPLOYMENT successful (20.141649303s)
2016/01/19 07:52:14 STAGE RESULT: 1.web.shipping-company.com:22 - execution of stage CODE_DEPLOYMENT successful (20.142226488s)
2016/01/19 07:52:15 STAGE SUCCESS: CODE_DEPLOYMENT
2016/01/19 07:52:15 STARTING STAGE: POST_DEPLOYMENT
2016/01/19 07:52:15 2.web.shipping-company.com:22 -- STARTING: echo 'restarting webserver'
2016/01/19 07:52:15 1.web.shipping-company.com:22 -- STARTING: echo 'restarting webserver'
2016/01/19 07:52:15 1.workers.shipping-company.com:22 -- STARTING: echo 'restarting workers'
2016/01/19 07:52:15 1.workers.shipping-company.com:22 -- STDOUT -- restarting workers
2016/01/19 07:52:15 1.workers.shipping-company.com:22 -- SUCCESS: "echo 'restarting workers'"
2016/01/19 07:52:15 1.web.shipping-company.com:22 -- STDOUT -- restarting webserver
2016/01/19 07:52:15 1.web.shipping-company.com:22 -- SUCCESS: "echo 'restarting webserver'"
2016/01/19 07:52:16 2.web.shipping-company.com:22 -- STDOUT -- restarting webserver
2016/01/19 07:52:17 2.web.shipping-company.com:22 -- SUCCESS: "echo 'restarting webserver'"
2016/01/19 07:52:18 STAGE RESULT: 1.workers.shipping-company.com:22 - execution of stage POST_DEPLOYMENT successful (100.106365ms)
2016/01/19 07:52:19 Could not find deployment with id 235, database is locked
2016/01/19 07:52:19 Could not find deployment with id 235, database is locked
2016/01/19 07:52:19 Could not find deployment with id 235, database is locked
2016/01/19 07:52:19 STAGE RESULT: 1.web.shipping-company.com:22 - execution of stage POST_DEPLOYMENT successful (141.476426ms)
2016/01/19 07:52:19 STAGE RESULT: 2.web.shipping-company.com:22 - execution of stage POST_DEPLOYMENT successful (151.186347ms)
2016/01/19 07:52:20 Could not find deployment with id 235, database is locked
2016/01/19 07:52:20 Could not find deployment with id 235, database is locked
2016/01/19 07:52:20 Could not find deployment with id 235, database is locked
2016/01/19 07:52:20 Could not find deployment with id 235, database is locked
2016/01/19 07:52:20 Could not find deployment with id 235, database is locked
2016/01/19 07:52:20 Could not find deployment with id 235, database is locked
2016/01/19 07:52:20 Could not find deployment with id 235, database is locked
2016/01/19 07:52:20 Could not find deployment with id 235, database is locked
2016/01/19 07:52:20 Could not find deployment with id 235, database is locked
2016/01/19 07:52:20 STAGE SUCCESS: POST_DEPLOYMENT
2016/01/19 07:52:21 Could not find deployment with id 235, database is locked
2016/01/19 07:52:21 DEPLOYMENT FINISHED: deployment_id=235
```

The interesting lines are these:

```
2016/01/19 07:52:20 Could not find deployment with id 235, database is locked
```

I confirmed that this log output comes from the webhook notifier by changing the log output.

This makes the webhooks feature pretty unreliable since a lot (look at the ratio of log entries vs. "could not find deployment" messages) of possible events are not acted upon.

I'm still not sure how to fix this issue since I don't know yet who holds the lock. Maybe we need to change the architecture of the webhooks feature to reduce the amount of queries? (This won't fix the issue itself, but reduce the probability of it arising: configure the webhook notifier in `main()` only if any target has webhooks configured, instead of checking on every log message)

**Comment by mrnugget** [comment]
So, here's some more information. After reading through some go-sqlite3 issues I think that _maybe_ the problem lies in the configuration of the database connection. Here are some interesting issues to read through:

* https://github.com/mattn/go-sqlite3/issues/39
* https://github.com/mattn/go-sqlite3/issues/145
* https://github.com/mattn/go-sqlite3/issues/148
* https://github.com/mattn/go-sqlite3/issues/209

Sqlite3 shouldn't hold a read-lock in our use case: while trying to load the deployment from the database while the deployment is running, nobody else is inserting rows into the deployments table. But we [have a listener that writes every log entry to the database](https://github.com/applikatoni/applikatoni/blob/1159d2664c88d6dcdaa9ff82721cdfaf9c1918dd/server/main.go#L148-L149). This listener only adds rows to the `log_entries` table. And I guess that this write operation could happen at pretty much the same time as our read-operation (reading the deployment).


With this change proposed in several of the issues mentioned above I can't reproduce the error message anymore:

```go
	db, err = sql.Open("sqlite3", fmt.Sprintf("file:%s?cache=shared&mode=rwc", *databasePath))
	if err != nil {
		log.Fatal("could not open sqlite3 database file", err)
	}
	defer db.Close()
```

I don't want to add this though without really understanding the problem at hand and how this "fix" solves it. And I put "fix" in quotes here, because this change results in a deployment not being set to `successful` (I guess due to another race condition?).

```
2016/01/19 08:20:36 Could not update deployment state. err=database table is locked: deployments
```
(Originates [here](https://github.com/applikatoni/applikatoni/blob/1159d2664c88d6dcdaa9ff82721cdfaf9c1918dd/server/handlers.go#L279-L280))

**Comment by Wind0r** [comment]
Uff. That's not good to hear :(
While testing this never happen to me. Also i never thought about getting database locks.
Maybe I am using a older/newer version of mattn/go-sqlite3?

**Comment by mrnugget** [comment]
Which version are you running on?

```
% cd $GOPATH/src/github.com/mattn/go-sqlite3 && git show --quiet HEAD
commit b808f01f660667b1510acf216aa2b9fe7f555ddb
Author: looi <looi@users.noreply.github.com>
Date:   Sun Sep 20 10:07:23 2015 -0600

    Add FTS4 unicode61 tokenizer support
```

**Comment by Wind0r** [comment]
I am currently not able to check the version, I will check the version this evening. Sorry for that :(
I think my last update is a month ago.

**Comment by mrnugget** [comment]
Take your time! And no need to be sorry :)

I just checked our production instance and it seems like the problem doesn't arise there. But, since this looks and smells like a race-condition, the fact that it _can_ arise it enough for me personally to classify it as a bug.

**Comment by Wind0r** [comment]
Oh. My version is older as i thought.
I read the last commits. I do not think there happened anything important.
```
commit 2513631704612107a1c8b1803fb8e6b3dda2230e
Merge: a755cdc 21637a6
Author: mattn <mattn.jp@gmail.com>
Date:   Tue Nov 3 22:05:16 2015 +0900

    Merge pull request #251 from larsmans/fixes

    Clean up tempfiles in tests
```

Maybe something with sqlite?
```
sqlite3 --version
3.8.10.2 2015-05-20 18:17:19 2ef4f3a5b1d1d0c4338f8243d40a2452cc1f7fe4

**Comment by mrnugget** [comment]
In the vagrant box where I compiled the Linux builds and on our production system:

```
vagrant@vagrant-ubuntu-trusty-64:~$ sqlite3 --version
3.8.2 2013-12-06 14:53:30 27392118af4c38c5203a04b8013e1afdb1cebd0d
```

And on OS X, my local setup (where I ran into this bug):

```
% sqlite3 --version
3.8.10.2 2015-05-20 18:17:19 2ef4f3a5b1d1d0c4338f8243d40a2452cc1f7fe4
```

**Comment by mrnugget** [comment]
Small update after a few days of investigation:

1. `db.SetMaxOpenConns(1)` seems to fix the problem by essentially wrapping every access to the database in a mutex. Surprisingly, this doesn't slow us down.
2. `sql.Open("sqlite3", "database_file.sqlite?cache=shared&mode=rwc")` doesn't fix the problem.
3. `sql.Open("sqlite3", "database_file.sqlite?cache=shared&mode=rwc&_busy_timeout=9999999")` _kinda_ fixes the problems, but makes everything really, really slow.

More information and a link to a gist I wrote that demonstrates the `database is locked` issue can [be found here](https://github.com/mattn/go-sqlite3/issues/274).

It seems to me like we have two options for Applikatoni:

1. `db.SetMaxOpenConns(1)` -- I'm not sure about the tradeoffs here yet.
2. Make a breaking change to the way webhooks work and trigger them a LOT less. The idea here is to use the new `DeploymentEventHub` and `Subscriber` system I converted the notifiers to (check out this commit https://github.com/applikatoni/applikatoni/commit/1286a6ccbad49d88093fd1484f73c45159818083 to see it.) Webhooks would then only be triggered whenever the state of a `Deployment` changes (which is at maximum 3 times). Since the current filters use the `LogEntryType` they wouldn't work anymore. My idea here is to change the webhook system so that webhooks get called for _every_ state change (remember: at maximum 3). The receiving webhook can then decide what to do with each event. The payload would also slightly change. As I said: a breaking change. But I think that it's, while not completely solving the race condition issue with database locks, a good solution in terms of system architecture.
